tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+
+        // Merge prologue and note script tree
+        let prologue_and_notes = CodeBlock::new_join([self.prologue.clone(), note_program_tree]);
+
+        // Merge prologue, note tree, tx script and epilogue
+        let program_root = CodeBlock::new_join([prologue_and_notes, tx_script_and_epilogue]);
+
+        // Create [CodeBlockTable] from [AssemblyContext]
+        let cb_table = self
+            .assembler
+            .build_cb_table(assembly_context)
+            .map_err(TransactionError::BuildCodeBlockTableFailed)?;
+
+        // Create transaction program
+        let program = Program::with_kernel(program_root, self.tx_kernel.clone(), cb_table);
+
+        // Create compiled transaction
+        // TODO: we always have a digest for a transaction script (noop code block root if no user
+        // tx scrip is provided) - should we we change the [CompiledTransaction] struct to take a
+        // digest instead of a Option<Digest> fro the tx script.
+        Ok(CompiledTransaction::new(account_id, notes, Some(tx_script.hash()), program))
+    }
+
+    /// Returns a [CodeBlock] which contains the note program tree.
+    fn create_note_program_tree(
+        &mut self,
+        account_id: &AccountId,
+        notes: &[Note],
+        assembly_context: &mut AssemblyContext,
+    ) -> Result<CodeBlock, TransactionError> {
+        // Fetch target account interface
+        let target_account_procs = self
+            .account_procedures
+            .get(account_id)
+            .ok_or(TransactionError::AccountInterfaceNotFound(*account_id))?;
+
+        // Create note programs
+        let mut note_program_tree = notes
+            .iter()
+            .map(|note| {
+                let note_root = self
+                    .assembler
+                    .compile_in_context(note.script().code().clone(), assembly_context)
+                    .map_err(|_| TransactionError::CompileNoteScriptFailed)?;
+                let note_branches = collect_call_branches(&note_root);
+                verify_note_account_compatibility(
+                    note_root.hash(),
+                    &note_branches,
+                    target_account_procs,
+                )?;
+                Ok(CodeBlock::new_join([self.note_setup.clone(), note_root]))
+            })
+            .collect::<Result<Vec<CodeBlock>, TransactionError>>()?;
+
+        // Merge the note programs into a tree using join blcoks
+        while note_program_tree.len() != 1 {
+            // Pad note programs to an even number using a [Operation::Noop] span block
+            if note_program_tree.len() % 2 != 0 {
+                note_program_tree.push(CodeBlock::new_span(vec![Operation::Noop]));
+            }
+
+            note_program_tree = note_program_tree
+                .chunks(2)
+                .map(|code_blocks| {
+                    CodeBlock::new_join([code_blocks[0].clone(), code_blocks[1].clone()])
```

Comment:
Lots of cloning takes place when constructing the note tree. Can we do better?

Reply:
I think we should be able to do this without cloning. This might require manipulating two separate vectors one of which gets consumed on each iteration.